csrc/device_lower/pass/circular_buffer.cpp

```diff
@@ -1148,9 +1148,38 @@ class IsCircularBufferLoadLoop : public kir::IrVisitor {
 // associated load expressions.
 class CircularBufferLoopNestInspector : private kir::IrVisitor {
  public:
-  static InsertionInfo run(const std::vector<Expr*>& exprs) {
+  static std::pair<InsertionInfo, InsertionInfo> run(
+      const std::vector<Expr*>& exprs) {
     CircularBufferLoopNestInspector inspector(exprs);
-    return inspector.insertion_info_;
+
+    // InsertionInfo holds all circular buffer for-loops and is ordered from
+    // inner-most to outer-most. Split it into warp specialized and pipeline
+    // circular buffers. Enforce that we can only nest pipeline circular
+    // buffering inside of warp-specialization.
+    InsertionInfo ws_info;
+    InsertionInfo pipeline_info;
+    for (auto&& [cb_loop, cb_exprs] : inspector.insertion_info_) {
+      bool use_warp_specialization = std::holds_alternative<WarpSpecialized>(
+          GpuLower::current()
+              ->circularBufferInfo()
+              .getCircularBufferOptionsFor(cb_loop->iter_domain())
+              .type);
+      if (use_warp_specialization) {
+        ws_info[cb_loop] = cb_exprs;
+      } else {
+        NVF_ERROR(
+            ws_info.empty(),
```

Comment:
>     // InsertionInfo holds all circular buffer for-loops and is ordered from
>     // inner-most to outer-most.

The comment says something like that, I'm wondering where are we enforcing that order in the iteration? The order seems to come naively from the range based loop of an unordered_map. I don't think this gives us any control on that.

Reply:
* Track depth of circular for-loop in `for_loop_` stack.
* Get warp-specialized circular loops first. Get inner-most depth for them.
* Get pipeline circular buffers. Check that every depth is greater than the innermost warp-specialized loop.